monkey/infection_monkey/network_scanning/mssql_fingerprinter.py

```diff
@@ -77,27 +80,45 @@ def _query_mssql_for_instance_data(host: str) -> Optional[bytes]:
         sock.close()
 
 
-def _get_services_from_server_data(data: bytes) -> Dict[str, Any]:
-    services = {MSSQL_SERVICE: {}}
-    services[MSSQL_SERVICE]["display_name"] = DISPLAY_NAME
-    services[MSSQL_SERVICE]["port"] = SQL_BROWSER_DEFAULT_PORT
+def _get_tcp_port_from_response(response: str) -> Optional[NetworkPort]:
+    response_info = response.split(";")
+    if len(response_info) > 1:
+        response_info_keys = response_info[::2]
+        response_info_values = response_info[1::2]
+
+        response_dict = dict(zip(response_info_keys, response_info_values))
+
+        return response_dict.get("tcp", None)
+
+    return None
+
+
+def _get_services_from_server_data(data: bytes) -> List[DiscoveredService]:
+    services = []
 
     # Loop through the server data
-    mssql_instances = filter(lambda i: i != "", data[3:].decode().split(";;"))
-
-    for instance in mssql_instances:
-        instance_info = instance.split(";")
-        if len(instance_info) > 1:
-            services[MSSQL_SERVICE][instance_info[1]] = {}
-            for i in range(1, len(instance_info), 2):
-                # Each instance's info is nested under its own name, if there are multiple
-                # instances
-                # each will appear under its own name
-                services[MSSQL_SERVICE][instance_info[1]][instance_info[i - 1]] = instance_info[i]
-
-            logger.debug(f"Found MSSQL instance: {instance}")
-
-    if len(services[MSSQL_SERVICE].keys()) == 2:
-        services = {}
+    # Example data:
+    # yServerName;MSSQL-16;InstanceName;MSSQLSERVER;IsClustered;No;Version;14.0.1000.169;tcp;1433;np;\\MSSQL-16\pipe\sql\query;;
+    mssql_responses = filter(lambda i: i != "", data[3:].decode().split(";;"))
+    for response in mssql_responses:
+
+        response_tcp_port = _get_tcp_port_from_response(response)
+        if response_tcp_port is not None:
+            services.append(
+                DiscoveredService(
+                    protocol=NetworkProtocol.TCP,
+                    port=NetworkPort(response_tcp_port),
+                    services=NetworkService.MSSQL,
+                )
+            )
+        else:
+            services.append(
+                DiscoveredService(
+                    protocol=NetworkProtocol.UDP,
+                    port=SQL_BROWSER_DEFAULT_PORT,
+                    services=NetworkService.UNKNOWN,
+                )
+            )
```

Comment:
Why is it "else"? We've parsed the response, the port is obviously open by now.

Reply:
What if we don't have tcp in the response, what protocol is that and what service is that ???